syntax = "proto3";

package group;

import "google/protobuf/empty.proto";

message AddStudentRequest {
    string token = 1;
    string group_id = 2;
}

message AddWordToUserRequest {
    string word = 1;
    string definition = 2;
    string group_id = 3;
    uint32 user_id = 4;
    string token = 5;
}

message AddWordToUserResponse {
    uint32 word_id = 1;
}

message CreateGroupRequest {
    string title = 1;
    string token = 2;
}

message DeleteGroupRequest {
    string token = 1;
    string group_id = 2;
}

message FindGroupRequest {
    string token = 1;
    string group_id = 2;
}

message FindGroupsTeacherRequest {
    string token = 1;
}

message FindGroupsStudentRequest {
    string token = 1;
}

message GroupResponse {
    uint32 user_id = 1;
    string group_id = 2;
    string title = 3;
    repeated uint32 students = 4;
}

message FindStudentRequest {
    string token = 1;
    uint32 student_id = 2;
    string group_id = 3;
}

message StudentResponse {
    string email = 1;
    string username = 2;
}

message FindTeacherRequest {
    string token = 1;
    string group_id = 2;
}

message TeacherResponse {
    uint32 teacher_id = 1;
    string email = 2;
    string username = 3;
}

message GetStatisticsRequest {
    uint32 student_id = 1;
    string group_id = 2;
    string token = 3;
}

message StatisticsResponse {
    string stat_id = 1;
    string group_id = 2;
    uint32 teacher_id = 3;
    uint32 student_id = 4;
    repeated uint32 words = 5;
}

message RemoveStudentRequest {
    string token = 1;
    string group_id = 2;
    uint32 user_id = 3;
}

service GroupService {
    rpc AddStudent(AddStudentRequest) returns (google.protobuf.Empty);
    rpc AddWordToUser(AddWordToUserRequest) returns (AddWordToUserResponse);
    rpc CreateGroup(CreateGroupRequest) returns (google.protobuf.Empty);
    rpc DeleteGroup(DeleteGroupRequest) returns (google.protobuf.Empty);
    rpc FindGroup(FindGroupRequest) returns (GroupResponse);
    rpc FindStudent(FindStudentRequest) returns (StudentResponse);
    rpc FindTeacher(FindTeacherRequest) returns (TeacherResponse);
    rpc FindGroupsTeacher(FindGroupsTeacherRequest) returns (stream GroupResponse);
    rpc FindGroupsStudent(FindGroupsStudentRequest) returns (stream GroupResponse);
    rpc GetStatistics(GetStatisticsRequest) returns (StatisticsResponse);
    rpc RemoveStudent(RemoveStudentRequest) returns (google.protobuf.Empty);
}
